_posts/2015-09-01-Cocoa-Architecture:-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,146 @@
+
+## Sprinkling Analytics
+
+We took some of the ideas for [Aspect oriented programming](http://albertodebortoli.github.io/blog/2014/03/25/an-aspect-oriented-approach-programming-to-ios-analytics/) with [Analytics](https://cocoapods.org/pods/ARAnalytics#user-content-aspect-oriented-dsl).
+
+Where we used to [have this](https://github.com/artsy/energy/blob/master/Classes/Controllers/Popovers/Add%20to%20Album/ARAddToAlbumViewController.m#L271-L282):
+
+``` objc
+@implementation ARAddToAlbumViewController
+
+- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    if (indexPath.row < [self.albums count]) {
+        Album *selectedAlbum = ((Album *)self.albums[indexPath.row]);
+        ARTickedTableViewCell *cell = (ARTickedTableViewCell *)[self.tableView cellForRowAtIndexPath:indexPath];
+
+        if ([cell isSelected]) {
+            [ARAnalytics event:ARRemoveFromAlbumEvent withProperties:@{
+                @"artworks" : @(self.artworks.count),
+                @"from" : [ARNavigationController pageID]
+            }];
+            [...]
+```
+
+We would instead build something [like this](https://github.com/artsy/eigen/blob/master/Artsy/App/ARAppDelegate+Analytics.m#L69):
+
+``` objc
+@implementation ARAppDelegate (Analytics)
+
+- (void)setupAnalytics
+{
+    ArtsyKeys *keys = [[ArtsyKeys alloc] init];
+	[...]
+    [ARAnalytics setupWithAnalytics: @{ [...] } configuration:
+    @{
+        ARAnalyticsTrackedEvents:
+            @[
+                @{
+                    ARAnalyticsClass: ARAddToAlbumViewController.class,
+                    ARAnalyticsDetails: @[
+                        @{
+                            ARAnalyticsEventName: ARRemoveFromAlbumEvent,
+                            ARAnalyticsSelectorName: NSStringFromSelector(@selector(tableView: didSelectRowAtIndexPath:)),
+                            ARAnalyticsProperties: ^NSDictionary*(ARAddToAlbumViewController *controller, NSArray *_) {
+                                return @{
+                                    @"artworks" : @(controller.artworks.count),
+                                    @"from" : [ARNavigationController pageID],
+                                };
+                        },
+                        [...]
+                    ]
+                },
+				[...]
+```
+
+This gives us the ability to not sprinkle analytics code around the app in every file. It keeps the responsibilities of objects simpler and we've been happy with it in Eigen. We've not migrated it into Energy, it's dependency on ReactiveCocoa brings too much additional weight. If you want to learn more about this pattern check out [Aspect-Oriented Programming and ARAnalytics](http://artsy.github.io/blog/2014/08/04/aspect-oriented-programming-and-aranalytics/).
```

Comment:
> it's dependency on ReactiveCocoa

No apostrophe. Also, could mention what/if we're planning on addressing that disadvantage with a move to something else, etc.